diesel/src/macros/mod.rs

```diff
@@ -26,47 +26,31 @@ macro_rules! __diesel_column {
         impl SelectableExpression<$($table)::*> for $column_name {
         }
 
-        impl AppearsOnTable<$($table)::*> for $column_name {
-        }
-
-        impl<Right, Kind> SelectableExpression<
-            Join<$($table)::*, Right, Kind>,
-        > for $column_name where
-            $column_name: AppearsOnTable<Join<$($table)::*, Right, Kind>>,
-        {
-        }
-
-        impl<Left> SelectableExpression<
-            Join<Left, $($table)::*, Inner>,
-        > for $column_name where
-            Left: $crate::JoinTo<$($table)::*>
+        impl<QS> AppearsOnTable<QS> for $column_name where
+            QS: AppearsInFromClause<$($table)::*, Count=Once>,
         {
         }
 
-        impl<Right, Kind> AppearsOnTable<
-            Join<$($table)::*, Right, Kind>,
+        impl<Left, Right> SelectableExpression<
+            Join<Left, Right, LeftOuter>,
         > for $column_name where
-            Right: QuerySource,
-            $($table)::*: $crate::JoinTo<Right>
+            $column_name: AppearsOnTable<Join<Left, Right, LeftOuter>>,
+            Left: AppearsInFromClause<$($table)::*, Count=Once>,
+            Right: AppearsInFromClause<$($table)::*, Count=Never>,
```

Comment:
ℹ️ Constraint B: Can only select column when column's table appears on left side of Left Outer Join and not on right side